File: linear_title_researching.py
#!/usr/bin/env python3
"""Update Linear issue titles when status changes to "to research"."""

from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request
from typing import Any

LINEAR_API_URL = "https://api.linear.app/graphql"
TITLE_MARKER = "Cursor researching"


def _normalize_status(value: str | None) -> str:
    if not value:
        return ""
    return " ".join(value.strip().lower().split())


def should_mark_issue(payload: dict[str, Any]) -> bool:
    """Return True when this webhook means the issue entered to research."""
    trigger_context = payload.get("triggerContext", {}) if isinstance(payload, dict) else {}
    new_status = trigger_context.get("newStatus")
    return _normalize_status(new_status) == "to research"


def build_updated_title(title: str, marker: str = TITLE_MARKER) -> str:
    """Prefix marker to title once, preserving existing marker."""
    cleaned_title = title.strip()
    if cleaned_title.lower().startswith(marker.lower()):
        return cleaned_title
    return f"{marker} {cleaned_title}"


def _graphql_request(api_key: str, query: str, variables: dict[str, Any]) -> dict[str, Any]:
    payload = json.dumps({"query": query, "variables": variables}).encode("utf-8")
    request = urllib.request.Request(
        LINEAR_API_URL,
        data=payload,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Authorization": api_key,
        },
    )
    with urllib.request.urlopen(request, timeout=30) as response:
        body = response.read().decode("utf-8")
    result = json.loads(body)
    if result.get("errors"):
        raise RuntimeError(f"Linear GraphQL error: {result['errors']}")
    return result


def _extract_issue_identifier(payload: dict[str, Any]) -> str:
    trigger_context = payload.get("triggerContext", {}) if isinstance(payload, dict) else {}
    identifier = trigger_context.get("id")
    if not isinstance(identifier, str) or not identifier.strip():
        raise ValueError("Could not read issue identifier from payload.triggerContext.id")
    return identifier.strip()


def _fetch_issue_by_identifier(api_key: str, identifier: str) -> dict[str, str] | None:
    query = """
    query FindIssueByIdentifier($identifier: String!) {
      issues(filter: { identifier: { eq: $identifier } }, first: 1) {
        nodes {
          id
          identifier
          title
        }
      }
    }
    """
    response = _graphql_request(api_key, query, {"identifier": identifier})
    nodes = response.get("data", {}).get("issues", {}).get("nodes", [])
    if not nodes:
        return None
    node = nodes[0]
    return {
        "id": node["id"],
        "identifier": node["identifier"],
        "title": node["title"],
    }


def _update_issue_title(api_key: str, issue_id: str, new_title: str) -> bool:
    mutation = """
    mutation UpdateIssueTitle($id: String!, $title: String!) {
      issueUpdate(id: $id, input: { title: $title }) {
        success
      }
    }
    """
    response = _graphql_request(api_key, mutation, {"id": issue_id, "title": new_title})
    return bool(response.get("data", {}).get("issueUpdate", {}).get("success"))


def run(payload: dict[str, Any], api_key: str) -> int:
    if not should_mark_issue(payload):
        print("Status is not 'to research'; no title update required.")
        return 0

    issue_identifier = _extract_issue_identifier(payload)
    issue = _fetch_issue_by_identifier(api_key, issue_identifier)
    if issue is None:
        print(f"Issue {issue_identifier} not found in Linear.")
        return 1

    new_title = build_updated_title(issue["title"])
    if new_title == issue["title"].strip():
        print(f"Issue {issue_identifier} already contains '{TITLE_MARKER}'.")
        return 0

    success = _update_issue_title(api_key, issue["id"], new_title)
    if not success:
        print(f"Failed to update title for issue {issue_identifier}.")
        return 1

    print(f"Updated issue {issue_identifier} title to: {new_title}")
    return 0


def _read_payload(path: str | None) -> dict[str, Any]:
    if path:
        with open(path, "r", encoding="utf-8") as handle:
            return json.load(handle)
    return json.load(sys.stdin)


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Add 'Cursor researching' to Linear issue title when status changes to 'to research'."
    )
    parser.add_argument(
        "--payload",
        default=None,
        help="Path to a JSON payload file (if omitted, reads from stdin).",
    )
    args = parser.parse_args()

    api_key = os.getenv("LINEAR_API_KEY")
    if not api_key:
        print("Missing LINEAR_API_KEY environment variable.")
        return 1

    try:
        payload = _read_payload(args.payload)
        return run(payload, api_key)
    except (ValueError, json.JSONDecodeError) as exc:
        print(f"Invalid payload: {exc}")
        return 1
    except urllib.error.URLError as exc:
        print(f"Network error while contacting Linear: {exc}")
        return 1
    except RuntimeError as exc:
        print(str(exc))
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/linear-title-researching.mjs
const CURSOR_RESEARCHING_PREFIX = 'Cursor researching';
const LINEAR_GRAPHQL_ENDPOINT = 'https://api.linear.app/graphql';

function compact(value) {
  return typeof value === 'string' ? value.trim() : '';
}

function readPath(source, path) {
  return path.reduce((value, key) => {
    if (value == null || typeof value !== 'object') {
      return undefined;
    }

    return value[key];
  }, source);
}

function firstStringAtPath(source, paths) {
  for (const path of paths) {
    const value = compact(readPath(source, path));
    if (value) {
      return value;
    }
  }

  return '';
}

export function statusNameFromPayload(payload) {
  return firstStringAtPath(payload, [
    ['triggerContext', 'newStatus'],
    ['triggerContext', 'status'],
    ['data', 'node', 'state', 'name'],
    ['data', 'issue', 'state', 'name'],
    ['data', 'state', 'name'],
    ['issue', 'state', 'name'],
    ['state', 'name']
  ]);
}

export function isStatusChangedToResearch(payload) {
  const triggerName = compact(payload?.triggerContext?.trigger).toLowerCase();
  const actionName = compact(payload?.action).toLowerCase();
  const webhookType = compact(payload?.triggerContext?.webhookType).toLowerCase();
  const issueType = compact(payload?.type).toLowerCase();
  const statusName = statusNameFromPayload(payload).toLowerCase();

  const isIssueEvent = webhookType === 'issue' || issueType === 'issue';
  const isStatusChange = triggerName === 'status_changed' || actionName === 'update';

  return isIssueEvent && isStatusChange && statusName === 'to research';
}

export function titleWithCursorResearching(title) {
  const currentTitle = compact(title);
  if (!currentTitle) {
    return CURSOR_RESEARCHING_PREFIX;
  }

  if (currentTitle.toLowerCase().includes(CURSOR_RESEARCHING_PREFIX.toLowerCase())) {
    return currentTitle;
  }

  return `${CURSOR_RESEARCHING_PREFIX}: ${currentTitle}`;
}

export function issueIdFromPayload(payload) {
  return firstStringAtPath(payload, [
    ['triggerContext', 'id'],
    ['data', 'node', 'id'],
    ['data', 'issue', 'id'],
    ['issue', 'id'],
    ['id']
  ]);
}

export function issueTitleFromPayload(payload) {
  return firstStringAtPath(payload, [
    ['triggerContext', 'title'],
    ['data', 'node', 'title'],
    ['data', 'issue', 'title'],
    ['issue', 'title'],
    ['title']
  ]);
}

export function buildIssueTitleMutation(issueId, title) {
  return {
    query: `
      mutation UpdateIssueTitle($id: String!, $title: String!) {
        issueUpdate(id: $id, input: { title: $title }) {
          success
          issue {
            id
            title
          }
        }
      }
    `,
    variables: {
      id: issueId,
      title
    }
  };
}

export async function updateIssueTitle({ issueId, title, apiKey, fetchImpl = globalThis.fetch }) {
  if (!apiKey) {
    throw new Error('LINEAR_API_KEY or LINEAR_API_TOKEN is required');
  }

  if (!issueId) {
    throw new Error('Linear issue id is required');
  }

  if (!title) {
    throw new Error('Linear issue title is required');
  }

  if (typeof fetchImpl !== 'function') {
    throw new Error('A fetch implementation is required');
  }

  const response = await fetchImpl(LINEAR_GRAPHQL_ENDPOINT, {
    method: 'POST',
    headers: {
      Authorization: apiKey,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(buildIssueTitleMutation(issueId, title))
  });

  const body = await response.json();
  if (!response.ok || body.errors?.length) {
    const errorMessage = body.errors?.map((error) => error.message).join('; ') || response.statusText;
    throw new Error(`Linear issue title update failed: ${errorMessage}`);
  }

  if (!body.data?.issueUpdate?.success) {
    throw new Error('Linear issue title update did not report success');
  }

  return body.data.issueUpdate.issue;
}

export async function handleLinearStatusChanged(payload, options = {}) {
  if (!isStatusChangedToResearch(payload)) {
    return {
      changed: false,
      reason: 'not_to_research_status_change'
    };
  }

  const issueId = issueIdFromPayload(payload);
  const nextTitle = titleWithCursorResearching(issueTitleFromPayload(payload));

  if (nextTitle === issueTitleFromPayload(payload)) {
    return {
      changed: false,
      reason: 'title_already_marked',
      issueId,
      title: nextTitle
    };
  }

  const issue = await updateIssueTitle({
    issueId,
    title: nextTitle,
    apiKey: options.apiKey ?? process.env.LINEAR_API_KEY ?? process.env.LINEAR_API_TOKEN,
    fetchImpl: options.fetchImpl
  });

  return {
    changed: true,
    issue
  };
}

async function readPayloadFromStdin() {
  const chunks = [];

  for await (const chunk of process.stdin) {
    chunks.push(chunk);
  }

  return Buffer.concat(chunks).toString('utf8');
}

async function readPayload() {
  const rawPayload =
    process.env.LINEAR_WEBHOOK_PAYLOAD ||
    process.env.CURSOR_AUTOMATION_TRIGGER ||
    process.env.TRIGGER_CONTEXT ||
    (await readPayloadFromStdin());

  if (!compact(rawPayload)) {
    throw new Error('No Linear webhook payload provided');
  }

  return JSON.parse(rawPayload);
}

export async function main() {
  const payload = await readPayload();
  const result = await handleLinearStatusChanged(payload);
  console.log(JSON.stringify(result, null, 2));
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
  });
}
